python/singa/sonnx.py

```diff
@@ -0,0 +1,242 @@
+        model = pickle.load(f)
+    return model
+
+def onnx_model_init(inputs,model):
+    '''
+    init onnx model
+    Args:
+        inputs: input data for model
+        model: onnx model
+    Return:
+        a dictionary for save all of the node of singa computation graph
+    '''
+    a = {}
+    a['X'] = inputs
+    for i in model.graph.node:
+        if (i.op_type == 'Constant'):
+            a[str(i.output[0])] = tensor.from_numpy(numpy_helper.to_array(i.attribute[0].t))
+            a[str(i.output[0])].stores_grad = True
+    return a
+
+def onnx_loss(a,model,target):
+    '''
+    get onnx model loss
+    Args:
+        a: singa computational graph nodes dictionary
+        model: onnx model
+        target:
+    Return:
+        loss for onnx model
+    '''
+    for i in model.graph.node:
+        if (i.op_type == 'Constant'):
+            pass
+            # do nothing
+        if (i.op_type == 'LeakyRelu'):
+            a[str(i.output[0])] = autograd.relu(a[str(i.input[0])])
+        elif (i.op_type == 'Softmax'):
+            a[str(i.output[0])] = autograd.softmax(a[str(i.input[0])])
+        elif (i.op_type == 'Add'):
+            if(str(i.input[1])[-1] == 'b'):
+                a[str(i.output[0])] = autograd.add_bias(a[str(i.input[0])], a[str(i.input[1])])
+            else:
+                a[str(i.output[0])] = autograd.add(a[str(i.input[0])],a[str(i.input[1])])
+        elif (i.op_type == 'MatMul'):
+            a[str(i.output[0])] = autograd.matmul(a[str(i.input[0])], a[str(i.input[1])])
+
+    loss = autograd.cross_entropy(a['Y'], target)
+    return loss
+
+
+
+def get_onnx_model(y, dy=None):
+    '''
+    from singa computational graph(loss) to get onnx model
+    Args:
+        y:loss of singa autograd
+    Return:
+        onnx model
+    '''
+    ######################
+    import onnx
+    from onnx import helper
+    from onnx import AttributeProto, TensorProto, GraphProto
+    from onnx import numpy_helper
+    import numpy as np
+    from singa import tensor
+    X = helper.make_tensor_value_info('X', TensorProto.FLOAT, [1, 2])
+    Y = helper.make_tensor_value_info('Y', TensorProto.FLOAT, [1, 2])
+    node = []
+    ######################
+
+    dependency = infer_dependency(y.creator)
+    #get singa computational graph dependency
+
+    assert y.size() == 1, 'y must be a Tensor with a single value;' \
+                          'size of y is % d' % y.size()
+
+    # by default the dy is a tensor with 1.0 for each sample;
+    if dy is None:
+        dy = float(1.0)
+    elif isinstance(dy, Tensor):
+        dy = dy.data
+    else:
+        dy = float(dy)
+
+    # ready is a queue of (operation, dy list)
+    ready = deque([(y.creator, (dy,))])
+    # using bfs iterative method to check all computational graph nodes
+    not_ready = {}  # mapping: op->[dy]
+    gradients = {}  # mapping: x->dx if x.stores_grad
+    if y.stores_grad:
+        gradients[y] = dy
+
+    while len(ready) > 0:
```

Comment:
this block of code can be simplified.  no need to do unnecessary backward calculations and store the calculated gradients.

Reply:
yes